src/halmos/utils.py

```diff
@@ -941,283 +949,75 @@ class EVM:
-    0x81630654DFB0FD282A37117995646CDDE2CF8EEFE9F3F96FDB12CFDA88DF6668: 245,
-    0xDDF78CFA378B5E068A248EDAF3ABEF23EA9E62C66F86F18CC5E695CD36C9809B: 246,
-    0xE9944EBEF6E5A24035A31A727E8FF6DA7C372D99949C1224483B857F6401E346: 247,
-    0x6120B123382F98F7EFE66ABE6A3A3445788A87E48D4E6991F37BAADCAC0BEF95: 248,
-    0x168C8166292B85070409830617E84BDD7E3518B38E5AC430DC35ED7D16B07A86: 249,
-    0xD84F57F3FFA76CC18982DA4353CC5991158EC5AE4F6A9109D1D7A0AE2CBA77ED: 250,
-    0x3E7257B7272BB46D49CD6019B04DDEE20DA7C0CB13F7C1EC3391291B2CCEBABC: 251,
-    0x371F36870D18F32A11FEA0F144B021C8B407BB50F8E0267C711123F454B963C0: 252,
-    0x9346AC6DD7DE6B96975FEC380D4D994C4C12E6A8897544F22915316CC6CCA280: 253,
-    0x54075DF80EC1AE6AC9100E1FD0EBF3246C17F5C933137AF392011F4C5F61513A: 254,
-    0xE08EC2AF2CFC251225E1968FD6CA21E4044F129BFFA95BAC3503BE8BDB30A367: 255,
-}
+class OffsetMap:
+    """A specialized mapping class that splits integer keys into two parts for efficient storage and lookup.
+
+    Keys are split into:
+    1. High-order bits used as the main dictionary key
+    2. Low-order bits (offset) stored with the value
+
+    For example, with offset_bits=16:
+    - Key 0x1234567890 is split into:
+      - High part: 0x123456 (dictionary key)
+      - Low part: 0x7890 (offset)
+
+    Useful for storing values where keys are related by small offsets.
+    """
 
-sha3_inv_offset: dict[int, tuple[int, int]] = mk_sha3_inv_offset(sha3_inv)
+    def __init__(self, offset_bits: int = 16):
+        """Initialize the OffsetMap.
+
+        Args:
+            offset_bits: Number of bits to use for the offset part of the key.
+                        The remaining bits are used as the main dictionary key.
+        """
+        self._map = {}
+        self._offset_bits = offset_bits
+        self._mask = (1 << offset_bits) - 1  # Creates a mask of offset_bits 1's
+
+    def __getitem__(self, k: int) -> tuple[int, int]:
+        """Get the value and offset delta for a key.
+
+        Args:
+            k: The integer key to look up
+
+        Returns:
+            A tuple of (value, delta) where:
+            - value is the stored value (or None if not found)
+            - delta is the difference between the requested offset and stored offset
+              Note: delta can be negative, e.g., when computing slot(a[n-1]) which is `(keccak(slot(a)) - 1) + n`
+        """
+        (v, offset) = self._map.get(k >> self._offset_bits, (None, None))
+        if v is None:
+            return (None, None)
+        delta = (k & self._mask) - offset
+        return (v, delta)
+
+    def __setitem__(self, k: int, v: Any):
+        """Store a value with its offset.
+
+        Args:
+            k: The integer key to store
+            v: The value to store
+        """
+        self._map[k >> self._offset_bits] = (v, k & self._mask)
+
+    def copy(self) -> "OffsetMap":
+        new_map = OffsetMap(self._offset_bits)
+        new_map._map = self._map.copy()
+        return new_map
```

Comment:
same as before, if this is expected to be append-only can we avoid copies?

Reply:
this copy() method is used when creating a branch.  if we just return self, any new items added after the branch creation would appear in both paths, which would break the current code logic.

that said, we might be able to improve the logic to avoid copies, e.g., by annotating map items with their originating path info. let's discuss this optimization possibility separately.